adapters/admatic/admatic.go

```diff
@@ -0,0 +1,155 @@
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &adapter{
+		endpoint: template,
+	}
+
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	Host := ""
+
+	for _, imp := range request.Imp {
+		admaticExt, err := getImpressionExt(imp)
+		if err != nil {
+			return nil, []error{err}
+		}
+		if admaticExt.Host != "" {
+			if Host == "" {
+				Host = admaticExt.Host
+			} else if Host != admaticExt.Host {
+				return nil, []error{&errortypes.BadInput{
+					Message: "There must be only one Hos",
+				}}
+			}
+		} else {
+			return nil, []error{&errortypes.BadInput{
+				Message: "The Hos must not be empty",
+			}}
+		}
+	}
+
+	resolvedUrl, err := a.resolveUrl(Host)
+	if err != nil {
+		return nil, []error{err}
+	}
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+	requestData := &adapters.RequestData{
+		Method: "POST",
+		Uri:    resolvedUrl,
+		Body:   requestJSON,
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func getImpressionExt(imp openrtb2.Imp) (*openrtb_ext.ImpExtAdmatic, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Bidder extension not provided or can't be unmarshalled",
+		}
+	}
+
+	var admaticExt openrtb_ext.ImpExtAdmatic
+	if err := json.Unmarshal(bidderExt.Bidder, &admaticExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error while unmarshaling bidder extension",
+		}
+	}
+
+	return &admaticExt, nil
+}
+
+// "Un-templates" the endpoint by replacing macroses and adding the required query parameters
+func (a *adapter) resolveUrl(host string) (string, error) {
+	params := macros.EndpointTemplateParams{Host: host}
+
+	endpointStr, err := macros.ResolveMacros(a.endpoint, params)
+	if err != nil {
+		return "", err
+	}
+
+	parsedUrl, err := url.Parse(endpointStr)
+	if err != nil {
+		return "", err
+	}
+
+	return parsedUrl.String(), nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur

```

Comment:
by default line 130 will initialize `bidResponse.Currency` as `USD`

reassigning `bidResponse.Currency` on line 131  without checking  `response.Cur` length may overwrite `bidResponse.Currency` with blank value

code block can be refactored as below

```
	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
        if len(response.Cur) != 0 {
	    bidResponse.Currency = response.Cur
        }
```